lib/internal/benchmark/clock.js

```diff
@@ -0,0 +1,195 @@
+  codes: { ERR_BENCHMARK_MISSING_OPERATION },
+} = require('internal/errors');
+const { validateInteger, validateNumber } = require('internal/validators');
+
+let debugBench = require('internal/util/debuglog').debuglog('benchmark', (fn) => {
+  debugBench = fn;
+});
+
+const kUnmanagedTimerResult = Symbol('kUnmanagedTimerResult');
+
+// If the smallest time measurement is 1ns
+// the minimum resolution of this timer is 0.5
+const MIN_RESOLUTION = 0.5;
+
+class Timer {
+  constructor() {
+    this.now = process.hrtime.bigint;
+  }
+
+  get scale() {
+    return 1e9;
+  }
+
+  get resolution() {
+    return 1 / 1e9;
+  }
+
+  /**
+   * @param {number} timeInNs
+   * @returns {string}
+   */
+  format(timeInNs) {
+    validateNumber(timeInNs, 'timeInNs', 0);
+
+    if (timeInNs > 1e9) {
+      return `${NumberPrototypeToFixed(timeInNs / 1e9, 2)}s`;
+    }
+
+    if (timeInNs > 1e6) {
+      return `${NumberPrototypeToFixed(timeInNs / 1e6, 2)}ms`;
+    }
+
+    if (timeInNs > 1e3) {
+      return `${NumberPrototypeToFixed(timeInNs / 1e3, 2)}us`;
+    }
+
+    if (timeInNs > 1e2) {
+      return `${NumberPrototypeToFixed(timeInNs, 0)}ns`;
+    }
+
+    return `${NumberPrototypeToFixed(timeInNs, 2)}ns`;
+  }
+}
+
+const timer = new Timer();
+
+class ManagedTimer {
+  /**
+   * @type {bigint}
+   */
+  #start;
+  /**
+   * @type {bigint}
+   */
+  #end;
+  /**
+   * @type {number}
+   */
+  #iterations;
+  /**
+   * @type {number}
+   */
+  #recommendedCount;
+
+  /**
+   * @param {number} recommendedCount
+   */
+  constructor(recommendedCount) {
+    this.#recommendedCount = recommendedCount;
+  }
+
+  /**
+   * Returns the recommended value to be used to benchmark your code
+   * @returns {number}
+   */
+  get count() {
+    return this.#recommendedCount;
+  }
+
+  /**
+   * Starts the timer
+   */
+  start() {
+    this.#start = timer.now();
+  }
+
+  /**
+   * Stops the timer
+   * @param {number} [iterations=1] The amount of iterations that run
+   */
+  end(iterations = 1) {
+    this.#end = timer.now();
+
+    validateInteger(iterations, 'iterations', 1);
+
+    this.#iterations = iterations;
+  }
+
+  [kUnmanagedTimerResult]() {
+    if (this.#start === undefined)
+      throw new ERR_BENCHMARK_MISSING_OPERATION('You forgot to call .start()');
+
+    if (this.#end === undefined)
+      throw new ERR_BENCHMARK_MISSING_OPERATION('You forgot to call .end(count)');
+
+    return [Number(this.#end - this.#start), this.#iterations];
+  }
+}
+
+function createRunUnmanagedBenchmark(awaitOrEmpty) {
+  return `
+const startedAt = timer.now();
+
+for (let i = 0; i < count; i++)
+  ${awaitOrEmpty}bench.fn();
+
+const duration = Number(timer.now() - startedAt);
+return [duration, count];
+`;
+}
+
+function createRunManagedBenchmark(awaitOrEmpty) {
+  return `
+${awaitOrEmpty}bench.fn(timer);
+
+return timer[kUnmanagedTimerResult]();
+  `;
+}
+
+const AsyncFunction = async function() { }.constructor;
+const SyncFunction = function() { }.constructor;
```

Comment:
I think we add those validation to the module `internal/validators` and so they can be used in another places -  For example:

```js
function validateAsyncFunction(value, name) {
  if (typeof value !== 'function') {
    throw new ERR_INVALID_ARG_TYPE(name, 'function', value);
  }

  return fn.constructor.name === 'AsyncFunction';
}

function validateSyncFunction(value, name) {
  if (typeof value !== 'function') {
    throw new ERR_INVALID_ARG_TYPE(name, 'function', value);
  }

  return fn.constructor.name === 'Function';
}

```

What do you think about this? @RafaelGSS

Reply:
That would work if we validate the parameter, but in our case we just need the `isAsync` call to check if `async function ()` was passed

Reply:
Just a question, - do you think there is a way to achieve it so that these validations can be used elsewhere?

If not, no worries 😁

Reply:
If we use it in different modules, yes. But I don't think that's the case.